### DISCUSSION TITLE
Generating custom classes

### DISCUSSION THREAD

**Discussion by waj00115** [tailwindlabs/tailwindcss#9254] [Help] [answered]
All,

Looking for some help on what to add to my tailwind.config.js file.

I'm working for a website that already has FontAwesome implemented with the duotone set, embedding with webfonts and CSS.

Their documentation lists to change the primary and secondary colors of an individual duotone icon to do the following:

`<i class="fa-duotone fa-camera" style="--fa-primary-color: #123456; --fa-secondary-color: #654321"></i>`

What I'd like to do is generate CSS classes called fa-color-primary-XXXXX-XXX and fa-color-secondary-XXXXX-XXX so that if I want the camera icon to be, say, blue with yellow, I can do:

`<i class="fa-duotone fa-camera fa-color-primary-blue-700 fa-color-secondary-yellow-300"></i>`

Where tailwind somewhere would have generated:
```
.fa-color-primary-blue-700 {
    --fa-primary-color: INSERT-blue-700-HERE;
}

.fa-color-secondary-yellow-300 {
  --fa-secondary-color: INSERT-yellow-300-HERE;
}
```

I'd like these generated not only as options for the built-in color palette, but I have custom colors as well. Only a single color for each, not a range running from 50 to 900.

I read the documentation and I think I want to create a plugin with matchUtilities, but I javascript isn't my forte and I'm not sure where to begin.

Any help would be greatly appreciated.

**Comment by wongjn** [comment, marked as answer]
Your plugin could look something like this:

```js
const flattenColorPalette = require('tailwindcss/lib/util/flattenColorPalette');
const plugin = require('tailwindcss/plugin');

module.exports = plugin(({ matchUtilities, theme }) => {
  matchUtilities(
    {
      // Key is the class name prefix, `fa-color-primary-*`.
      // `value` is the color value.
      'fa-color-primary': (value) => ({
        // The CSS declaration.
        '--fa-primary-color': value,
      }),
      'fa-color-secondary': (value) => ({
        '--fa-secondary-color': value,
      }),
    },
    {
      // Let Tailwind know what CSS value types we accept (for arbitrary variant usage).
      type: ['color'],
      // List of values to generate candidate classes from, taken from the resolved values
      // of `theme.colors`. It is the resolved values so colors in user's Tailwind config
      // will be included too.
      values: flattenColorPalette(theme('colors')),
    },
  );
});
```
[Here it is working in a Tailwind Play](https://play.tailwindcss.com/DN3hzxG0mA?file=config).

**Reply by waj00115** [reply]
Thank you -- I was in the neighborhood but would have never thought to use flattenColorPalette. It certainly makes since as I look at the code of the function that you'd need to convert the colors from a multi-dimensional array into a single-dimensional array.

I see it's working in Tailwind Play, but when I created a new tailwind project to try this out to make sure nothing was hanging around from my config that was goofing it up, I'm getting a **TypeError: flattenColorPalette is not a function**.

I dropped what you had in Play into a config file as-is (only added a couple custom theme colors. Any ideas?

Thank you so much!

**Reply by wongjn** [reply]
You might need to alter it like so:
```diff
-const flattenColorPalette = require('tailwindcss/lib/util/flattenColorPalette');
+const { default: flattenColorPalette } = require('tailwindcss/lib/util/flattenColorPalette');
```
If that doesn't work, then it must be something else within your environment.

**Reply by waj00115** [reply]
That fixed it. Thank you so much for all of your help, and your incredibly speedy followup!
